keras/callbacks.py

```diff
@@ -823,38 +828,113 @@ def is_indexed_slices(grad):
+                        x, y, sample_weight = self.model._standardize_user_data(
+                            x, y, sample_weight=sample_weight,
+                            batch_size=batch_size)
+                        batch_val = x + y + sample_weight
+
+                        if (self.model.uses_learning_phase and
+                                not isinstance(K.learning_phase(), int)):
+                            batch_val += [0.]
+                        if not len(batch_val) == len(tensors):
+                            raise ValueError('validation_data generator\'s output '
+                                             'length ({0}) must be the same as '
+                                             'the sum of the lengths of the '
+                                             'model\'s inputs, targets and '
+                                             'sample_weights ({1}).'
+                                             .format(len(batch_val), len(tensors)))
+
+                        feed_dict = dict(zip(tensors, batch_val))
+                        result = self.sess.run([self.merged], feed_dict=feed_dict)
+                        summary_str = result[0]
+                        self.writer.add_summary(summary_str, epoch)
+                        i += 1
+                else:
+                    val_data = self.validation_data
+                    tensors = (self.model.inputs +
+                               self.model.targets +
+                               self.model.sample_weights)
 
-                val_data = self.validation_data
-                tensors = (self.model.inputs +
-                           self.model.targets +
-                           self.model.sample_weights)
-
-                if self.model.uses_learning_phase:
-                    tensors += [K.learning_phase()]
-
-                assert len(val_data) == len(tensors)
-                val_size = val_data[0].shape[0]
-                i = 0
-                while i < val_size:
-                    step = min(self.batch_size, val_size - i)
                     if self.model.uses_learning_phase:
-                        # do not slice the learning phase
-                        batch_val = [x[i:i + step] for x in val_data[:-1]]
-                        batch_val.append(val_data[-1])
-                    else:
-                        batch_val = [x[i:i + step] for x in val_data]
-                    assert len(batch_val) == len(tensors)
-                    feed_dict = dict(zip(tensors, batch_val))
-                    result = self.sess.run([self.merged], feed_dict=feed_dict)
-                    summary_str = result[0]
-                    self.writer.add_summary(summary_str, epoch)
-                    i += self.batch_size
+                        tensors += [K.learning_phase()]
+                    if not len(val_data) == len(tensors):
+                        raise ValueError('validation_data length ({0}) must be '
```

Comment:
Please add a note saying that the learning_phase should be added by `fit_generator`, otherwise it's a little confusing.

Reply:
If I understood you correctly, fit_generator shouldn't add the learning_phase, this is why we add it at line 894:
`batch_val += [0.]`
I copied this behaviour from function `predict_on_batch` in training.py that is used in regular validation generator evaluation.

Reply:
This is the non-generator path. 
0 is added to val_data here : https://github.com/keras-team/keras/pull/9796/files/86e5448ff06d30bebfdf8a4781562ad6abaabd1c#diff-b25d82c3f751f73f6e62b8455547ac73R124

Reply:
Done